### PR TITLE
Add reserved keywords

### PR DESCRIPTION
**Pull request by gregli-msft**
Adding reserved keywords, to avoid a conflict in the future.  Done at the lexer level, similar to how the other keywords are handled.  Most of the code here is just about plumbing down the flag.

Resolves https://github.com/microsoft/Power-Fx/issues/1344

### CODE REVIEW

**Comment by lesaltzm** [comment]
1. Is this necessary?
2. Why not just a reserved identifier? Something in the symbol table, perhaps.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 1]
```diff
@@ -82,7 +85,20 @@ public enum Flags
 
         // Pretty Print defaults
         public const string FourSpaces = "    ";
-        public const string LineBreakAndfourSpaces = "\n    ";
+        public const string LineBreakAndfourSpaces = "\n    ";
+
+        // Reserved Keywords
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:90]: change comment to: Reserved but currently unused keywords

File: src/libraries/Microsoft.PowerFx.Core/Parser/ParserOptions.cs [hunk 2]
```diff
@@ -25,6 +25,11 @@ public class ParserOptions
         /// </summary>
         public bool NumberIsFloat { get; set; }
 
+        /// <summary>
+        /// If true, various words have been reserved and are not available for identifiers.
+        /// </summary>
+        public bool ReservedKeywords { get; set; }
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/ParserOptions.cs:31]: This should not be public on ParserOptions, imo. Everyone other than PA should always get the reserved behavior.

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Parser/ParserOptions.cs:31]: probably internal works

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ReservedKeywordPositive.txt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+>> {Children:1}
+Errors: Error 1-9: Unexpected characters. The formula contains 'Error' where 'Ident' is expected.|Error 1-9: Unexpected characters. The formula contains 'Error' where 'Colon' is expected.|Error 1-9: Expected colon. We expect a colon (:) at this point in the formula.|Error 9-10: Unexpected characters. The formula contains 'Colon' where 'CurlyClose' is expected.|Error 9-10: Unexpected characters. Characters are used in the formula in an unexpected way.
+
+>> {Siblings:1}
+Errors: Error 1-9: Unexpected characters. The formula contains 'Error' where 'Ident' is expected.|Error 1-9: Unexpected characters. The formula contains 'Error' where 'Colon' is expected.|Error 1-9: Expected colon. We expect a colon (:) at this point in the formula.|Error 9-10: Unexpected characters. The formula contains 'Colon' where 'CurlyClose' is expected.|Error 9-10: Unexpected characters. Characters are used in the formula in an unexpected way.
+
+>> {'OkToUse' :1}
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ReservedKeywordPositive.txt:44]: also test `X Is Y`, should result in the reserved error

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ReservedKeywordPositive.txt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+// With ReservedKeywords enabled (PowerFxV1) these tests should all fail.
+
+>> {OkToUse :1}
+{OkToUse:1}
+
+>> {As :1}
+Errors: Error 1-3: Unexpected characters. The formula contains 'As' where 'Ident' is expected.|Error 1-3: Unexpected characters. The formula contains 'As' where 'Colon' is expected.|Error 1-3: Expected colon. We expect a colon (:) at this point in the formula.|Error 4-5: Unexpected characters. The formula contains 'Colon' where 'CurlyClose' is expected.|Error 4-5: Unexpected characters. Characters are used in the formula in an unexpected way.
+
+>> {Is      :1}
+Errors: Error 1-3: Unexpected characters. The formula contains 'Error' where 'Ident' is expected.|Error 1-3: Unexpected characters. The formula contains 'Error' where 'Colon' is expected.|Error 1-3: Expected colon. We expect a colon (:) at this point in the formula.|Error 9-10: Unexpected characters. The formula contains 'Colon' where 'CurlyClose' is expected.|Error 9-10: Unexpected characters. Characters are used in the formula in an unexpected way.
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ReservedKeywordPositive.txt:12]: Shouldn't these errors contain your new error message?

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs (hunk 1 above)
Comment (gregli-msft) [reply src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:90]: Done.

File: src/libraries/Microsoft.PowerFx.Core/Parser/ParserOptions.cs (hunk 2 above)
Comment (gregli-msft) [reply src/libraries/Microsoft.PowerFx.Core/Parser/ParserOptions.cs:31]: Done.

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs [hunk 5]
```diff
@@ -179,7 +179,7 @@ internal static bool TryGetBestOverload(CheckTypesContext context, IErrorContain
                         // and, trivially, the nodeToCoercedTypeMap are properly set at this point.
                         bestOverload = maybeFunc;
                         return true;
-                    }
+                    }
```
[the remark this answers is not in the document]
Comment (gregli-msft) [reply src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs]: Done.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ReservedKeywordPositive.txt (hunk 3 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ReservedKeywordPositive.txt:44]: Added.  Getting odd results and not seeing my error message.  I'll need to dig into this.

**Comment by gregli-msft** [comment]
Linking with https://github.com/microsoft/Power-Fx/issues/1385

**Review by lesaltzm — APPROVED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs [hunk 6]
```diff
@@ -164,6 +166,8 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
             var parameterSymbols = runtimeConfig?.Values?.SymbolTable;
             var symbolsAll = ReadOnlySymbolTable.Compose(parameterSymbols, symbolTable);
 
+            options ??= new ParserOptions() { ReservedKeywords = Config.Features.ReservedKeywords };
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:169]: this is the wrong place.  It hides nulls.   Can you update Engine.GetDefaultParserOptionsCopy()?

Comment (gregli-msft) [reply src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs]: Thanks, done!  GetDefaultParserOptionsCopy() has already been updated.

**Review by MikeStall — APPROVED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**